traversal/scanner/GraphIterator.java

```diff
@@ -206,28 +206,51 @@ private void initialiseStart() {
     private void initialiseEnds() {
         assert toTraverse.isEmpty();
         toRevisit.addAll(procedure.endVertices());
-        direction = Direction.REVISIT;
+        direction = Direction.REVISIT_RETRIEVED;
     }
 
     private boolean computeAnswer() {
         while ((direction == Direction.TRAVERSE && !toTraverse.isEmpty()) ||
-                (direction == Direction.REVISIT && !toRevisit.isEmpty())) {
+                ((direction == Direction.REVISIT_ALL || direction == Direction.REVISIT_RETRIEVED) && !toRevisit.isEmpty())) {
             ProcedureVertex<?, ?> vertex;
             if (direction == Direction.TRAVERSE) {
                 toTraverse.remove(vertex = toTraverse.first());
                 traverse(vertex);
-            } else {
+            } else if (direction == Direction.REVISIT_ALL) {
                 toRevisit.remove(vertex = toRevisit.last());
                 revisit(vertex);
+            } else {
+                toRevisit.remove(vertex = toRevisit.last());
+                revisitRetrieved(vertex);
```

Comment:
either we are in the revisiting any vertex state, or in the revisiting retrieved vertices only state